source/lib/cache.js

```diff
@@ -1,97 +1,147 @@
-	const dataToStore = {
-		dateCached: new Date().toUTCString(),
-		timeToCache: cacheTime,
-		value: value,
+	return {
+		value: JSON.parse(value),
+		source,
+		...annotateCacheEntry({ts, ttl}),
 	}
-	return AsyncStorage.setItem(`aao:${key}`, JSON.stringify(dataToStore))
-}
-function getItem(key: string): CacheResultType<any> {
-	return AsyncStorage.getItem(`aao:${key}`).then(stored =>
-		annotateCacheEntry(JSON.parse(stored)),
-	)
 }
 
-/// MARK: Help tools
+global._setItem = setItem
+global._getItem = getItem
+global._fetchAndCacheItem = fetchAndCacheItem
 
-const helpToolsKey = 'help:tools'
-const helpToolsCacheTime = [1, 'hour']
-import {type ToolOptions} from '../views/help/types'
-const {data: helpData} = require('../../docs/help.json')
-export function setHelpTools(tools: Array<ToolOptions>) {
-	return setItem(helpToolsKey, tools, helpToolsCacheTime)
-}
-export function getHelpTools(): CacheResultType<?Array<ToolOptions>> {
-	return getItem(helpToolsKey)
-}
-function fetchHelpToolsBundled(): Promise<Array<ToolOptions>> {
-	return Promise.resolve(helpData)
-}
-function fetchHelpToolsRemote(): Promise<{data: Array<ToolOptions>}> {
-	return fetchJson(API('/tools/help'))
-}
-export async function fetchHelpTools(): Promise<Array<ToolOptions>> {
-	const cachedValue = await getHelpTools()
+export async function fetchAndCacheItem(args: {
+	key: string,
+	url?: string,
+	cbForBundledData?: () => Promise<mixed>,
+	ttl?: [number, string],
+	afterFetch?: (parsed: any) => any,
+	force?: boolean,
+	delay?: boolean,
+}): CacheResult<any> {
+	let {
+		key,
+		url,
+		cbForBundledData,
+		ttl,
+		afterFetch,
+		force: shouldForce,
+		delay: shouldDelay,
+	} = args
+
+	let start = Date.now()
 
-	if (process.env.NODE_ENV === 'development') {
-		return fetchHelpToolsBundled()
+	console.log('fetchAndCacheItem', {key, url, cbForBundledData, ttl})
+
+	// if (process.env.NODE_ENV === 'development' && cbForBundledData) {
+	// 	let data = await cbForBundledData()
+	// 	await setItem(key, data, {source: url})
+	// 	return getItem(key)
+	// }
+
+	let {value, isExpired, isCached, source} = await getItem(key)
+
+	if (!isExpired && value != null && !shouldForce) {
+		return {value, isCached, isExpired, source}
 	}
 
-	if (!cachedValue.isExpired && cachedValue.value) {
-		return cachedValue.value
+	if (!source && url) {
+		source = url
 	}
 
 	try {
-		const request = await fetchHelpToolsRemote()
-		await setHelpTools(request.data)
+		// try fetching first
+		let body = await fetchJson(url)
+
+		if (afterFetch) {
+			body = afterFetch(body)
+		}
+
+		await setItem(key, body, {source: url, ttl})
 
-		return request.data
+		let elapsed = Date.now() - start
+		if (shouldDelay && elapsed < 500) {
+			// 0.5s delay for ListViews – if we let them go at full speed, it feels broken
+			await delay(500 - elapsed)
+		}
+
+		return getItem(key)
 	} catch (error) {
-		if (cachedValue.isCached && cachedValue.value) {
-			return cachedValue.value
+		// fall back to the local data
```

Comment:
This comment doesn't seem to match the code that follows it, is this meant for the "catch"?

Reply:
Removed.